DatadogTrace/Sources/DatadogTracer.swift

```diff
@@ -101,7 +101,9 @@ internal class DatadogTracer: OTTracer {
 
     func extract(reader: OTFormatReader) -> OTSpanContext? {
         // TODO: RUMM-385 - make `HTTPHeadersReader` available in public API
-        reader.extract()
+        var reader = reader as? TracePropagationHeadersReader
+        reader?.tracerSampleRate = localTraceSampler.samplingRate
+        return (reader as? OTFormatReader)?.extract()
```

Comment:
Fair call 👍. Like you said @maxep, head-based sampling requires touching multiple moving parts and I found it impossible fit it in seamlessly. This piece is a clear stain, let's work on it 🙌.

I like the 2nd proposal more as it is better scoped. The `DDSpanContext` is already a `struct`, so it is cheap to create it again. Also, in next PR we will likely deprecate HTTP header writers and readers as being not backward-compatible for the 2 remaining scenarios outlined in the PR description. `TracePropagationHeadersReader` may become obsolete in consequence.

Keep in mind tho, that 2nd approach will include a trap. There will be no good default we could extract out of headers for non-optional `DDSpanContext.sampleRate`, so we'd need to pass it a "whatever" value that will be overwritten later in `Tracer`:
```swift
extension TracePropagationHeadersReader where Self: OTFormatReader {
    public func extract() -> OTSpanContext? {
        guard let ids = read() else {
            return nil
        }
        return DDSpanContext(
            traceID: ids.traceID,
            spanID: ids.spanID,
            parentSpanID: ids.parentSpanID,
            baggageItems: BaggageItems(),
            sampleRate: 0, // [!!!] placeholder, it will be overwritten in Tracer.extract()
            isKept: sampled ?? false // unreachable default: we got trace ID, so this request must have been instrumented
        )
    }
}
```

By principle, **this situation should be solved by making `sampleRate` an optional member of `DDSpanContext`**. This however moves another part → the `DatadogInternal.TraceContext` that will also require optionality in `sampleRate`. This begins the chain reaction and `guard let { /* not reachable */ }` surgery in network instrumentation code. All that for the sake of principle → not worth it 🙅. I think we should accept `0` default as it has no practical consequence (it will be always overwritten by `Tracer`).

Reply:
That would make sense 👍 We could still require an optional sample rate in the reader, such as:
```swift
public protocol TracePropagationHeadersReader {
    /// The sample rate used to sample this trace.
    var tracerSampleRate: Float? { get } // getter only that can be set at init

    func read() -> (
        traceID: TraceID,
        spanID: SpanID,
        parentSpanID: SpanID?,
        sampled: Bool? // can this be part of the read method?
    )?
}
```
But as you said, if `0` as default as no consequences lets remove the sampling rate from the reader 👍 Especially if the current writer/reader become deprecated!

Reply:
✅ done